### PR TITLE
Enforce strides to be not null when ndim is nonzero

### PR DESCRIPTION
**Pull request by tqchen**
This PR updates the spec to enforce the strides to be not null when the ndim is nonzero.
This is going to help consumers to handle cases more uniformly.

### CODE REVIEW

File: include/dlpack/dlpack.h [hunk 1]
```diff
@@ -257,8 +257,9 @@ typedef struct {
   /*! \brief The shape of the tensor */
   int64_t* shape;
   /*!
-   * \brief strides of the tensor (in number of elements, not bytes)
-   *  can be NULL, indicating tensor is compact and row-majored.
+   * \brief strides of the tensor (in number of elements, not bytes),
+   *  can not be NULL if ndim != 0, must points to
+   *  an array of ndim elements that specifies the strides.
```
Comment (hpkfft) [inline include/dlpack/dlpack.h]: I suggest **requiring** `NULL` if `ndim == 0`.   That is:
```
 *  which is NULL if and only if ndim == 0, pointing to
 *  an array of ndim elements that specify the strides.
```

**Review by junrushao — APPROVED** [review]

File: include/dlpack/dlpack.h [hunk 2]
```diff
@@ -257,8 +257,9 @@ typedef struct {
   /*! \brief The shape of the tensor */
```
Comment (hpkfft) [inline include/dlpack/dlpack.h]: I suggest **requiring** NULL if `ndim == 0`.  That is:
```
 *  \brief The shape of the tensor, which is NULL if and only if ndim == 0,
 *   pointing to an array of ndim elements.
```

**Review by hpkfft — DISMISSED** [review]

File: include/dlpack/dlpack.h (hunk 1 above)
Comment (seberg) [reply include/dlpack/dlpack.h]: While I do think suggesting it makes sense, I am not sure I see a benefit (how can you rely on it)?

On a general note, we really can't just change this without leaving a comment?  From a purity point of view: before version 1.2 the strides were allowed to be `NULL` when the array was C contiguous.
I'll settle for something fuzzier, but I feel we need a paper trail (for at least a year or so)?

It's not like we know nobody does this!

Comment (tqchen) [reply include/dlpack/dlpack.h]: let us go with suggesting it instead of requiring it for now. Agree on @seberg to add the paper trail here

**Comment by tqchen** [comment]
Updated to add the suggestion to set pointers to NULL when ndim==0, add notes about change in DLPack v1.2

**Comment by tqchen** [comment]
https://github.com/dmlc/dlpack/issues/177

File: include/dlpack/dlpack.h (hunk 1 above)
Comment (hpkfft) [reply include/dlpack/dlpack.h]: What do you mean by "it"?
I do not think it makes sense to *suggest* strides be NULL if ndim == 0 if you're going to *require* strides not to be NULL when ndim > 0.
Either suggest both (and then there is no need to increment the version number), or require both (in v1.2).

File: include/dlpack/dlpack.h [hunk 3]
```diff
@@ -254,11 +254,23 @@ typedef struct {
...
+   * \brief strides of the tensor (in number of elements, not bytes),
+   *  can not be NULL if ndim != 0, must points to
+   *  an array of ndim elements that specifies the strides.
+   *
+   *  When ndim == 0, we suggest to set the strides to NULL.
+   *
+   *  \note Before DLPack v1.2, strides can be NULL to indicate contiguous data.
+   *        This is not allowed in DLPack v1.2 and later. The rationale
+   *        is to simplify the consumer hanlding, and most frameworks already
+   *        always returns the strides when ndim != 0.
```
Comment (hpkfft) [inline include/dlpack/dlpack.h]: Minor: "handling" is misspelled.
Do you have evidence that most frameworks already always return non-NULL strides when ndim != 0?
NumPy, for example, returns NULL for C-contiguous arrays before today's not-yet-released development version.
Even if the statement is true for most frameworks, it may not be useful to mention it in the header since consumers ought to correctly handle NULL anyway (e.g., to support the current and past releases of NumPy).

File: include/dlpack/dlpack.h (hunk 1 above)
Comment (tqchen) [reply include/dlpack/dlpack.h]: I meant the ndim == 0 indicate strides == NULL. Like @seberg said, I agree it makes sense, however, also not sure see a benefit where one can rely on it(unlike the strides not null case), so would be good to know what are potential usecases here

Comment (seberg) [reply include/dlpack/dlpack.h]: I just don't see the reason to prescribe it. Forcing null seems like a code quality thing to to make sure things crash if incorrectly used and things aren't "undefined". Or do you have one?

File: include/dlpack/dlpack.h (hunk 3 above)
Comment (tqchen) [reply include/dlpack/dlpack.h]: noted, updated the typo. And rmeoved the note on frameworks here

File: include/dlpack/dlpack.h (hunk 1 above)
Comment (hpkfft) [reply include/dlpack/dlpack.h]: #177 seemed (originally, at least) to be motivated by memory safety (and, I would add, debugability).
In the case `ndim == 0`, either `strides` is NULL, or it is some garbage value (e.g., it points to unrelated allocated memory or it points to unallocated memory).   Well, I guess it could point to a new array of one element, but that seems unnatural and wasteful.   My question is, in the case `ndim == 0`, if `strides` is not NULL what other value would be reasonable/expected?

Comment (tqchen) [reply include/dlpack/dlpack.h]: Some implementation might do the following thing:

- Allocate DLTensor, plus the strides/shape inplace that follows the DLTensor
- `dltensor.shape =  reinterpret_cast<int64_t*>(reinterpret_cast<char*>(self) + sizeof(DLTensor))`
- `dltensor.strides =  dltensor.shape + dltensor.ndim`

This is actually quite convenient implementation and it is nice that it does not have to check for ndim = 0 and specially set shape/strides to NULL.

And because ndim == 0, it is safe in a sense that there is no memory value bein accessed from shape/strides

Comment (seberg) [reply include/dlpack/dlpack.h]: I would frame it as: the consumer can rely on `strides[dim]` working. Apparently that was desired.
To me that isn't much about memory safety, but maybe you can convince me that there is a serious memory safety angle somewhere additionally.

Comment (hpkfft) [reply include/dlpack/dlpack.h]: So, yeah, fundamentally just a code quality thing.
But I do think the clause "which is NULL if and only if ndim == 0" is clearer to humans than "cannot be NULL if ndim != 0".
From a verification point of view, it's easier to write a test suite when a standard specifies everything rather than when it leaves the value of `strides` unspecified in corner cases.
Just my two cents--I'm not trying to make trouble if others prefer to leave it unspecified.

Comment (hpkfft) [reply include/dlpack/dlpack.h]: If `ndim == 1`, then (in the future) the consumer can rely on `strides[0]` working.  Yes, I see that.

@seberg, in NumPy https://github.com/numpy/numpy/pull/29872, we now have:
```
  dl_tensor->shape = (ndim > 0) ? (int64_t *)((char *)ptr + offset) : NULL;
  dl_tensor->strides = (ndim > 0) ? dl_tensor->shape + ndim : NULL;
```
That seemed to me to be worth the minor inconvenience.
Otherwise, `shape` and `strides` would point past the end of allocated memory.
Is that still OK with you?

@tqchen, in the opening comment of #177, you wrote, "Enforcing strides to be not NULL would help with memory safety here. "   Would you be willing to further explain your motivation, and perhaps, what you feel is best from a memory safety point of view if `ndim ==0`?

Comment (tqchen) [reply include/dlpack/dlpack.h]: i somewhat agree with @seberg that consumers can rely on `strides[dim]` working for dim that `0<= dim < ndim`

Comment (tqchen) [reply include/dlpack/dlpack.h]: specificially, i think it is OK for `shape` and `strides`(and maybe desirable from the implementation simplicity pov) to point to the end of allocated memory for ndim=0 case

**Comment by tqchen** [comment]
Thanks everyone for the feedbacks. We updated to explicit state the constraint 
`consumer can always rely on strides[dim] being valid for 0 <= dim < ndim.` 

This ensures that the standard still allows shape/strides to points to the last of the allocated memory when ndim=0, and not having to special case for the scenario

**Comment by tqchen** [comment]
most outstanding comments are resolved, see also note above on the rationale on the ndim=0 case, i think we can plan to merge soon soon and have a 1.2 release. Thanks @hpkfft @seberg and everyone for chiming in

**Review by Kathryn-cat — APPROVED** [review]

**Comment by tqchen** [comment]
planning to merge in 24hours
